### PR TITLE
Email service

### PR DESCRIPTION
**Pull request by blacdev**

### CODE REVIEW

File: fundli/database/account.py [hunk 1]
```diff
@@ -18,6 +18,7 @@ def find_one(self, query_filter):
                 password=data.get("password"),
                 first_name=data.get("first_name"),
                 last_name=data.get("last_name"),
+                verification_code=data.get("verification_code"),
```
Comment (thealamu) [inline fundli/database/account.py:21]: `verification_code` is vague, we could need a verification code for password-reset, email-reset, and so on

File: fundli/services/account.py [hunk 2]
```diff
@@ -125,3 +128,89 @@ def is_valid_email(email):
...
+    if error:
+        return None, error
+    if not account:
+        return None, Error("account not found", 404)
+
+    _, error = verify_code(account, code)
+    if error:
+        return None, error
+
+    update_query = {"password": hash_password(password)}
```
Comment (thealamu) [inline fundli/services/account.py:195]: Can we do a full account update/replace?

**Review by thealamu — CHANGES_REQUESTED** [review]

File: fundli/account.py [hunk 3]
```diff
@@ -170,3 +173,74 @@ async def get_me(activeAccount: Account = Depends(authenticate)):
                     }
     """
     return account_response_serializer(activeAccount)
+
+
+@account_router.get("/forgot-password", response_model=EmailResponse)
```
Comment (thealamu) [inline fundli/account.py:178]: Should this be a POST?

File: fundli/services/account.py [hunk 4]
```diff
@@ -125,3 +128,90 @@ def is_valid_email(email):
...
+    if error:
+        return None, error
+    if not account:
+        return None, Error("account not found", 404)
+
+    _, error = verify_code(account, code)
+    if error:
+        return None, error
+
+    account_update = {"password": hash_password(password)}
```
Comment (thealamu) [inline fundli/services/account.py:196]: Similar to other service functions that do an update, let's pass the account down from the router, update the specific field and do an entire update using `find_one_and_replace`

**Review by thealamu — CHANGES_REQUESTED** [review]

File: fundli/database/database.py [hunk 5]
```diff
@@ -39,3 +39,6 @@ def find_paginated(self, query_filter, page, per_page):
         paginator.set_prev_page()
         paginator.set_next_page()
         return PaginatedResult(paginator, cur)
+
+    def update(self, query_filter, data):
+        self.collection.update_one(query_filter, {"$set": data})
```
[the remark this answers is not in the document]
Comment (blacdev) [reply fundli/database/database.py:44]: No but i have renamed the method to `fine_one_and_replace` . IT function remains the same. I am also changing the the implementation to take the object as suggested

File: fundli/account.py (hunk 3 above)
Comment (blacdev) [reply fundli/account.py:178]: Yes. It has also been fixed

File: fundli/services/account.py [hunk 6]
```diff
@@ -125,3 +128,89 @@ def is_valid_email(email):
...
+    range_start = 10 ** (n - 1)
+    range_end = (10**n) - 1
+    return randint(range_start, range_end)
+
+
+def forgot_password(email: str, code: int):
+    if not is_valid_email(email):
+        return None, Error("invalid email", 400)
+
+    account, error = get_account(email)
```
Comment (thealamu) [inline fundli/services/account.py:143]: Update looks good but can we receive the account from the route as opposed to fetching it here. Single Responsibility Principle

File: fundli/database/database.py [hunk 7]
```diff
@@ -39,3 +39,6 @@ def find_paginated(self, query_filter, page, per_page):
         paginator.set_prev_page()
         paginator.set_next_page()
         return PaginatedResult(paginator, cur)
+
+    def find_one_and_replace(self, query_filter, data):
+        self.collection.update_one(query_filter, {"$set": data})
```
Comment (thealamu) [inline fundli/database/database.py:44]: This call should be a find_one_and_replace for the setup to work

File: fundli/services/account.py [hunk 8]
```diff
@@ -125,3 +128,89 @@ def is_valid_email(email):
...
+
+    _, error = verify_code(account, code)
+    if error:
+        return None, error
+
+    account.password = hash_password(password)
+    query = {"_id": account.id}
+
+    try:
+        account_database.find_one_and_replace(query, account.to_dict())
```
Comment (thealamu) [inline fundli/services/account.py:199]: So we can keep the database method name as “update”, it just does a find_one_and_replace internally

**Review by thealamu — CHANGES_REQUESTED** [review]

File: fundli/database/database.py (hunk 7 above)
Comment (blacdev) [reply fundli/database/database.py:44]: It works. I already tested it. Maybe more context would help me see what is wrong here

File: fundli/services/account.py (hunk 8 above)
Comment (blacdev) [reply fundli/services/account.py:199]: now you lost me

File: fundli/services/account.py [hunk 9]
```diff
@@ -125,3 +129,72 @@ def is_valid_email(email):
...
+    old_account_obj = account
+
+    _, error = verify_code(account, code)
+    if error:
+        return None, error
+
+    account.password = hash_password(password)
+
+    try:
+        account_database.update(old_account_obj.to_dict(), account.to_dict())
```
Comment (thealamu) [inline fundli/services/account.py:183]: Please use a queryfilter that targets the account's id as opposed to the entire `old_account_obj` object here

**Review by thealamu — APPROVED** [review]

**Pull request merged**